writer.go

```diff
@@ -117,78 +101,86 @@ type WriterResponseEmitter struct {
 	emitted bool
 }
 
-func (re *WriterResponseEmitter) SetEncoder(mkEnc func(io.Writer) Encoder) {
+func (re *writerResponseEmitter) SetEncoder(mkEnc func(io.Writer) Encoder) {
 	re.enc = mkEnc(re.w)
 }
 
-func (re *WriterResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
-	err := re.Emit(&cmdkit.Error{Message: fmt.Sprint(v), Code: errType})
-	if err != nil {
-		panic(err)
+func (re *writerResponseEmitter) CloseWithError(err error) error {
+	cwe, ok := re.c.(interface{ CloseWithError(error) error })
+	if ok {
+		return cwe.CloseWithError(err)
+	}
+
+	if err == nil || err == io.EOF {
+		return re.Close()
 	}
+
+	return errors.New("provided closer does not support CloseWithError")
 }
 
-func (re *WriterResponseEmitter) SetLength(length uint64) {
+func (re *writerResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
+	err := &cmdkit.Error{Message: fmt.Sprint(v), Code: errType}
+	log.Errorf("error setting error %q: cannot set error with writerResponseEmitter", err)
```

Comment:
If we can't do the right thing here, maybe we should just remove this from the interface. We really don't want the user to try calling this and then return a "success" to the user.

Reply:
Actually, *hasn't* this been removed from the interface?

Reply:
Yes, looks like I forgot deleting it. done.